### DISCUSSION TITLE
Query regarding LLVM IR generation

### DISCUSSION THREAD

**Discussion by garadeaniket** [AdaptiveCpp/AdaptiveCpp#1303] [Q&A]
Could somebody explain  in depth , how the sycl code has been transformed to llvm IR using adaptivecpp

**Comment by illuhad** [comment]
This is a very broad question to answer "in detail" in a github discussion :-)

For our generic single-pass compiler (`--acpp-targets=generic`), I would refer you to my paper which describes in detail how it works:
https://dl.acm.org/doi/10.1145/3585341.3585351
(you can find a publicly available version here: https://cdrdv2-public.intel.com/786536/Heidelberg_IWOCL__SYCLCon_2023_paper_2566-1.pdf )

The other compilation flows rely on integration with other toolchains, and as such, the IR generation works the same as in those other compilation flows. For example, for `--acpp-targets=cuda`, the same happens as in the clang CUDA toolchain, for `--acpp-targets=hip`, the same happens as in the clang HIP toolchain, and `--acpp-targets=omp` effectively does a regular host compilation using OpenMP.

If you have a more specific question around IR generation, I can give you a more specific answer.

**Comment by garadeaniket** [comment]
Is above required transformation passes , if so can you mention what are they?

**Reply by illuhad** [reply]
The generic compiler runs a custom LLVM pass, `TargetSeparationPass` which detects, extracts and embeds LLVM IR for kernels. That pass in turn also runs some dead code elimination passes as an implementation detail.

At JIT time, additional custom passes are run to make the IR compatible with the conventions of various LLVM GPU backends. Again, all of this is already described in detail in the paper.
